src/shared/cache-types.cts

```diff
@@ -1,58 +1,133 @@
+  IncrementalCachedAppPageValue,
+  'kind'
+> & {
+  kind: 'APP_PAGE'
+}
+
+/**
+ * Used for storing in blobs and reading from blobs
+ */
+export type NetlifyCachedAppPageValue = Omit<
+  IncrementalCachedAppPageValueForMultipleVersions,
+  'rscData'
+> & {
   // Next.js stores rscData as buffer, while we store it as base64 encoded string
   rscData: string | undefined
-  revalidate?: Parameters<IncrementalCache['set']>[2]['revalidate']
+  revalidate?: Parameters<CacheHandler['set']>[2]['revalidate']
 }
 
-type CachedPageValue = Extract<IncrementalCacheValue, { kind: 'PAGE' }>
-
-export type NetlifyCachedPageValue = CachedPageValue & {
-  revalidate?: Parameters<IncrementalCache['set']>[2]['revalidate']
+/**
+ * Shape of the cache value that is returned from CacheHandler.get or passed to CacheHandler.set
+ */
+type IncrementalCachedPageValueForMultipleVersions = Omit<IncrementalCachedPageValue, 'kind'> & {
+  kind: 'PAGE' | 'PAGES'
 }
 
-export type CachedFetchValue = Extract<IncrementalCacheValue, { kind: 'FETCH' }>
+/**
+ * Used for storing in blobs and reading from blobs
+ */
+export type NetlifyCachedPageValue = IncrementalCachedPageValueForMultipleVersions & {
+  revalidate?: Parameters<CacheHandler['set']>[2]['revalidate']
+}
 
-export type NetlifyIncrementalCacheValue =
-  | Exclude<
-      IncrementalCacheValue,
-      CachedRouteValue | CachedPageValue | IncrementalCachedAppPageValue
-    >
-  | NetlifyCachedRouteValue
-  | NetlifyCachedPageValue
-  | NetlifyCachedAppPageValue
+export type CachedFetchValueForMultipleVersions = Omit<CachedFetchValue, 'kind'> & {
+  kind: 'FETCH'
+}
 
 type CachedRouteValueToNetlify<T> = T extends CachedRouteValue
   ? NetlifyCachedRouteValue
-  : T extends CachedPageValue
+  : T extends IncrementalCachedPageValue
     ? NetlifyCachedPageValue
     : T extends IncrementalCachedAppPageValue
       ? NetlifyCachedAppPageValue
       : T
 
 type MapCachedRouteValueToNetlify<T> = { [K in keyof T]: CachedRouteValueToNetlify<T[K]> }
 
+/**
+ * Used for storing in blobs and reading from blobs
+ */
 export type NetlifyCacheHandlerValue = MapCachedRouteValueToNetlify<CacheHandlerValue>
+
+/**
+ * Used for storing in blobs and reading from blobs
+ */
+export type NetlifyIncrementalCacheValue = NetlifyCacheHandlerValue['value']
+
+// type IncrementalCacheValueToMultipleVersions<T> = T extends CachedRouteValue ?
+type IncrementalCacheValueToMultipleVersions<T> = T extends CachedRouteValue
+  ? CachedRouteValueForMultipleVersions
+  : T extends IncrementalCachedPageValue
+    ? IncrementalCachedPageValueForMultipleVersions
+    : T extends IncrementalCachedAppPageValue
+      ? IncrementalCachedAppPageValueForMultipleVersions
+      : T extends CachedFetchValue
+        ? CachedFetchValueForMultipleVersions
+        : T extends CacheHandlerValue
+          ? {
+              [K in keyof T]: IncrementalCacheValueToMultipleVersions<T[K]>
+            }
+          : T
```

Comment:
🙈🙈🙈

Reply:
Yeah, this whole thing is quite a mess. It doesn't help that there are 2 type variants generally created there:
 1. (new)mapping types from currently used `next` version we have for types to not that version specific but instead have kind of union that "work" for all the next versions we support (which is what this inline comment highlights)
 2. Mapping those types to how we actually store data in blobs (i.e. instead of using `Buffer` we use `string` for storage or adding extra information about `revalidate` that is used to figure out cacheable route handler response maxage).

For 1. I did try having multiple `next` version in deps and import types from there and have some kind of union on them, but setup for it was pretty messy:
 - I could not use npm alias and have multiple next versions there because their react(-dom) peerDeps could not be satisfied at the same time)
 - Using unions there was still not enough because next@canary now uses const enum so I would have to do messy things either way.

So I don't think we should treat those types necessarily as safeguard that we are using them correctly for Next purposes (unfortunately :( ).

The option where this could maybe work is if we would have actual 2 (or more) variants of `CacheHandler` implementations, but because those would be largely the same we would either have almost copies of them or maybe some abstraction that would be more difficult to read :/